File: Buoy/BUOYListener.m
//  The MIT License (MIT)
//
//  Copyright (c) 2014 Intermark Interactive
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy of
//  this software and associated documentation files (the "Software"), to deal in
//  the Software without restriction, including without limitation the rights to
//  use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
//  the Software, and to permit persons to whom the Software is furnished to do so,
//  subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
//  FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
//  COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
//  IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
//  CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

#import "BUOYListener.h"

// Constant Strings
NSString * const kBUOYBeaconRangeIdentifier = @"com.BUOYBeacon.Region";
NSString * const kBUOYDidFindBeaconNotification = @"kBUOYDidFindBeaconNotification";
NSString * const kBUOYBeacon = @"kBUOYBeacon";
NSTimeInterval const kBUOYDefaultTimeInterval = 5;


// Interface
@interface BUOYListener() <CLLocationManagerDelegate>
@property (nonatomic, strong) CLLocationManager *locationManager;
@property (nonatomic, strong) NSArray *beaconRegions;

// Listening
@property (nonatomic) BOOL isListening;
@property (nonatomic) NSTimeInterval notificationInterval;
@property (nonatomic, strong) NSMutableDictionary *seenBeacons;
@end


// Implementation
@implementation BUOYListener

#pragma mark - Singleton
+ (instancetype)defaultListener {
	static id _sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sharedInstance = [[self alloc] init];
    });
    
    return _sharedInstance;
}


#pragma mark - Init
- (instancetype)init {
    if (self = [super init]) {
        self.locationManager = [[CLLocationManager alloc] init];
        [self.locationManager setDelegate:self];
        [self.locationManager setDesiredAccuracy:kCLLocationAccuracyBest];
        self.notificationInterval = kBUOYDefaultTimeInterval;
    }
    
    return self;
}


#pragma mark - Start/Stop Listening
- (void)listenForBeaconsWithProximityUUIDs:(NSArray *)proximityIds {
    // Register for region monitoring
    NSMutableArray *beaconRegions = [NSMutableArray array];
    for (NSUUID *proximityId in proximityIds) {
        // Create the beacon region to be monitored.
        CLBeaconRegion *beaconRegion = [[CLBeaconRegion alloc] initWithProximityUUID:proximityId identifier:kBUOYBeaconRangeIdentifier];
        
        // Register the beacon region with the location manager.
        [self.locationManager startMonitoringForRegion:beaconRegion];
        [beaconRegions addObject:beaconRegion];
    }
    
    self.beaconRegions = beaconRegions;
}

- (void)listenForBeaconsWithProximityUUIDs:(NSArray *)proximityIds notificationInterval:(NSTimeInterval)seconds {
    self.notificationInterval = seconds;
    [self listenForBeaconsWithProximityUUIDs:proximityIds];
}

- (void)stopListening {
    for (CLBeaconRegion *region in self.beaconRegions) {
        [self.locationManager stopMonitoringForRegion:region];
    }
}

- (void)stopListeningForBeaconsWithProximityUUID:(NSUUID *)uuid {
    for (CLBeaconRegion *region in self.beaconRegions) {
        if ([region.proximityUUID isEqual:uuid]) {
            [self.locationManager stopMonitoringForRegion:region];
            break;
        }
    }
}


#pragma mark - Notifications
- (void)setNotificationInterval:(NSTimeInterval)seconds {
    self.notificationInterval = seconds;
}

- (void)sendNotificationWithBeacon:(CLBeacon *)beacon {
    if ([self shouldSendNotificationForBeacon:beacon]) {
        [[NSNotificationCenter defaultCenter] postNotificationName:kBUOYDidFindBeaconNotification object:nil userInfo:@{kBUOYBeacon:beacon}];
    }
}

- (BOOL)shouldSendNotificationForBeacon:(CLBeacon *)beacon {
    if (self.seenBeacons[[self keyForBeacon:beacon]]) {
        return abs([[NSDate date] timeIntervalSinceDate:self.seenBeacons[[self keyForBeacon:beacon]]]) > self.notificationInterval;
    }
    
    return YES;
}

- (NSString *)keyForBeacon:(CLBeacon *)beacon {
    return [NSString stringWithFormat:@"Buoy:%@:%@:%@", beacon.proximityUUID, beacon.major, beacon.minor];
}

- (void)addBeaconToSeenBeaconsDictionary:(CLBeacon *)beacon {
    [self.seenBeacons setObject:[NSDate date] forKey:[self keyForBeacon:beacon]];
}


#pragma mark - Location Manager Delegate
- (void)locationManager:(CLLocationManager *)manager didRangeBeacons:(NSArray *)beacons inRegion:(CLBeaconRegion *)region {
    // Notify for each Beacon found
    for (NSInteger b = 0; b < beacons.count; b++) {
        [self sendNotificationWithBeacon:beacons[b]];
    }
}

- (void)locationManager:(CLLocationManager *)manager didEnterRegion:(CLRegion *)region {
    if ([region isKindOfClass:[CLBeaconRegion class]]) {
        [self.locationManager startRangingBeaconsInRegion:(CLBeaconRegion *)region];
    }
}

@end

File: Buoy/Swift/BUOYListener.swift
//  The MIT License (MIT)
//
//  Copyright (c) 2014 Intermark Interactive
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy of
//  this software and associated documentation files (the "Software"), to deal in
//  the Software without restriction, including without limitation the rights to
//  use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
//  the Software, and to permit persons to whom the Software is furnished to do so,
//  subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
//  FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
//  COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
//  IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
//  CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import Foundation
import CoreLocation

let kBUOYRegionIdentifier = "com.BUOY.region.identifier"
let kBUOYDidFindBeaconNotification = "kBUOYDidFindBeaconNotification"
let kBUOYBeacon = "kBUOYBeacon"
var listener: BUOYListener? = nil

class BUOYListener: NSObject, CLLocationManagerDelegate {
    // Properties
    var manager: CLLocationManager
    var beaconRegions: [String : CLBeaconRegion]
    var interval: NSTimeInterval? = 0
    var seenBeacons: [String:NSDate]
    

    // Singleton Listener
    class func defaultListener() -> BUOYListener {
        if (listener == nil) {
            listener = BUOYListener()
        }
        return listener!
    }
    

    // Init
    override init(){
        self.manager = CLLocationManager()
        self.beaconRegions = Dictionary()
        self.seenBeacons = Dictionary()
        super.init()
        self.manager.delegate = self
        self.manager.requestAlwaysAuthorization()
    }
    
    
    // Start Listening
    func listenForBeacons(uuids: [NSUUID]) {
        for uuid in uuids {
            var region = CLBeaconRegion(proximityUUID: uuid, identifier: "\(kBUOYRegionIdentifier)-\(uuid)")
            region.notifyEntryStateOnDisplay = true
            self.manager.startMonitoringForRegion(region)
            self.manager.requestStateForRegion(region)
            self.beaconRegions[uuid.UUIDString] = region
        }
    }
    
    func listenForBeacons(uuids: [NSUUID], interval: NSTimeInterval) {
        self.interval = interval
        self.listenForBeacons(uuids)
    }
    
    
    // Stop Listening
    func stopListening() {
        for (uuid, region) in self.beaconRegions {
            self.manager.stopRangingBeaconsInRegion(region)
            self.manager.stopMonitoringForRegion(region)
            self.beaconRegions[uuid] = nil
        }
    }
    
    func stopListening(uuid: String) {
        if let beaconRegion = self.beaconRegions[uuid] {
            self.manager.stopRangingBeaconsInRegion(beaconRegion)
            self.manager.stopMonitoringForRegion(beaconRegion)
            self.beaconRegions[uuid] = nil
        }
    }
    
    
    // Notifications
    func setNotificationInterval(interval: NSTimeInterval) {
        self.interval = interval
    }
    
    func sendNotification(beacon: CLBeacon) {
        if self.shouldSendNotification(beacon) {
            self.addBeaconToSeenBeacons(beacon)
            NSNotificationCenter.defaultCenter().postNotificationName(kBUOYDidFindBeaconNotification, object:nil, userInfo:[kBUOYBeacon:beacon])
        }
    }
    
    func shouldSendNotification(beacon: CLBeacon) -> Bool {
        if let date = self.seenBeacons[beacon.buoyIdentifier()] {
            return abs(date.timeIntervalSinceNow) >= self.interval
        }
        
        return true
    }
    
    func addBeaconToSeenBeacons(beacon: CLBeacon) {
        self.seenBeacons[beacon.buoyIdentifier()] = NSDate()
    }
    
    
    // Location Manager Delegate
    func locationManager(manager: CLLocationManager!, didRangeBeacons beacons: [AnyObject]!, inRegion region: CLBeaconRegion!) {
        for beacon : AnyObject in beacons {
            if beacon is CLBeacon {
                self.sendNotification(beacon as! CLBeacon)
            }
        }
    }
    
    func locationManager(manager: CLLocationManager!, didDetermineState state: CLRegionState, forRegion region: CLRegion!) {
        if region.isKindOfClass(CLBeaconRegion.self) && state == CLRegionState.Inside {
            self.manager.startRangingBeaconsInRegion(region as! CLBeaconRegion)
        }
        else if region.isKindOfClass(CLBeaconRegion.self) && state == CLRegionState.Outside {
            self.manager.stopRangingBeaconsInRegion(region as! CLBeaconRegion)
        }
    }
}